src/Storages/VirtualColumnUtils.cpp

```diff
@@ -214,37 +215,63 @@ static void addPathAndFileToVirtualColumns(Block & block, const String & path, s
         block.getByName("_file").column->assumeMutableRef().insert(file);
     }
 
+    for (const auto & item : partitioning_keys)
+    {
+        if (block.has(item.name))
+        {
+            auto column = block.getByName(item.name).column;
+            ReadBufferFromString buf(item.column->getDataAt(0).toView());
+            item.type->getDefaultSerialization()->deserializeWholeText(column->assumeMutableRef(), buf, getFormatSettings(context));
+        }
+    }
+
     block.getByName("_idx").column->assumeMutableRef().insert(idx);
 }
 
-std::optional<ActionsDAG> createPathAndFileFilterDAG(const ActionsDAG::Node * predicate, const NamesAndTypesList & virtual_columns)
+std::optional<ActionsDAG> createPathAndFileFilterDAG(const ActionsDAG::Node * predicate, const NamesAndTypesList & virtual_columns, const String & path, const ContextPtr & context)
 {
     if (!predicate || virtual_columns.empty())
         return {};
 
     Block block;
+    std::unordered_map<std::string, std::string> keys;
+    if (context->getSettingsRef().use_hive_partitioning)
+        keys = parseHivePartitioningKeysAndValues(path);
     for (const auto & column : virtual_columns)
     {
-        if (column.name == "_file" || column.name == "_path")
+        if (column.name == "_file" || column.name == "_path" || keys.contains(column.name))
             block.insert({column.type->createColumn(), column.type, column.name});
     }
 
     block.insert({ColumnUInt64::create(), std::make_shared<DataTypeUInt64>(), "_idx"});
     return splitFilterDagForAllowedInputs(predicate, &block);
 }
 
-ColumnPtr getFilterByPathAndFileIndexes(const std::vector<String> & paths, const ExpressionActionsPtr & actions, const NamesAndTypesList & virtual_columns)
+ColumnPtr getFilterByPathAndFileIndexes(const std::vector<String> & paths, const ExpressionActionsPtr & actions, const NamesAndTypesList & virtual_columns, const ContextPtr & context)
 {
     Block block;
+    std::unordered_map<std::string, std::string> keys;
+    ColumnsWithTypeAndName partitioning_columns;
+    if (context->getSettingsRef().use_hive_partitioning)
+        keys = parseHivePartitioningKeysAndValues(paths[0]);
```

Comment:
Right now we use partition values only from the first path in the list of paths to filter all paths. And it's incorrect. Here we don't need it, we should fill `block` with all virtual columns that were created for hive partitioning (see my previous comments).
We should extract partition values and insert them into the block in `addFilterDataToVirtualColumns` separately for each path